lib/dynamo/dynamo/jobs.py

```diff
@@ -115,23 +115,39 @@ def _prepare_job_for_database(
     return prepared_job
 
 
-def _get_credit_cost(job: dict, costs: list[dict]) -> Decimal:
+def _get_credit_cost(job: dict, costs: dict) -> Decimal:
     job_type = job['job_type']
-    for cost_definition in costs:
-        if cost_definition['job_type'] == job_type:
-            if cost_definition.keys() not in ({'job_type', 'cost_parameter', 'cost_table'}, {'job_type', 'cost'}):
-                raise ValueError(f'Cost definition for job type {job_type} has invalid keys: {cost_definition.keys()}')
-
-            if 'cost_parameter' in cost_definition:
-                cost_parameter = cost_definition['cost_parameter']
-                parameter_value = job['job_parameters'][cost_parameter]
-                for item in cost_definition['cost_table']:
-                    if item['parameter_value'] == parameter_value:
-                        return item['cost']
-                raise ValueError(f'Cost not found for job type {job_type} with {cost_parameter} == {parameter_value}')
-
-            return cost_definition['cost']
-    raise ValueError(f'Cost not found for job type {job_type}')
+    cost_definition = costs[job_type]
+
+    if cost_definition.keys() not in ({'cost_parameters', 'cost_table'}, {'cost'}):
+        raise ValueError(f'Cost definition for job type {job_type} has invalid keys: {cost_definition.keys()}')
+
+    if 'cost' in cost_definition:
+        return cost_definition['cost']
+    else:
+        return _get_cost_from_table(job, cost_definition)
+
+
+def _get_cost_from_table(job: dict, cost_definition: dict) -> Decimal:
+    cost_lookup = cost_definition['cost_table']
+
+    for cost_parameter in cost_definition['cost_parameters']:
+        parameter_value = job['job_parameters'][cost_parameter]
+
+        if isinstance(parameter_value, list):
+            parameter_value = len(parameter_value)
+
+        if isinstance(parameter_value, float):
+            parameter_value = int(parameter_value)
+
+        try:
+            cost_lookup = cost_lookup[str(parameter_value)]
```

Comment:
I'd suggest factoring out a `_get_cost_parameter_value` helper function to help keep the logic for serializing parameter values self-contained:

```python
def _get_cost_parameter_value(job: dict, cost_parameter: str) -> str:
    parameter_value = job['job_parameters'][cost_parameter]

    if isinstance(parameter_value, list):
        parameter_value = len(parameter_value)

    if isinstance(parameter_value, float):
        parameter_value = int(parameter_value)
    
    return str(parameter_value)
```